godot-core/src/builtin/dictionary.rs

```diff
@@ -320,6 +317,232 @@ impl Share for Dictionary {
     }
 }
 
+struct DictionaryIterator<'a> {
+    last_key: Option<Variant>,
+    dictionary: &'a Dictionary,
+    start: bool,
+}
+
+impl<'a> DictionaryIterator<'a> {
+    fn new(dictionary: &'a Dictionary) -> Self {
+        Self {
+            last_key: None,
+            dictionary,
+            start: true,
+        }
+    }
+
+    fn call_init(dictionary: &Dictionary) -> Option<Variant> {
+        // SAFETY:
+        // `dictionary` is a valid `Dictionary` since we have a reference to it,
+        //    so this will call the implementation for dictionaries.
+        // `variant` is an initialized and valid `Variant`.
+        let variant: Variant = Variant::nil();
+        unsafe { Self::call_iter_fn(interface_fn!(variant_iter_init), dictionary, variant) }
+    }
+
+    fn call_next(dictionary: &Dictionary, last_var: Variant) -> Option<Variant> {
+        // SAFETY:
+        // `dictionary` is a valid `Dictionary` since we have a reference to it,
+        //    so this will call the implementation for dictionaries.
+        // `last_var` is an initialized and valid `Variant`, since we own a copy of it.
+        unsafe { Self::call_iter_fn(interface_fn!(variant_iter_next), dictionary, last_var) }
+    }
+
+    /// SAFETY:
+    /// It must be safe to call `iter_fn` using the provided `dictionary`, `next_var`
+    /// and a pointer to a `bool`, in addition `iter_fn` must return a valid `bool`.
+    unsafe fn call_iter_fn(
+        iter_fn: unsafe extern "C" fn(*const TagVariant, *mut TagVariant, *mut u8) -> u8,
+        dictionary: &Dictionary,
+        next_var: Variant,
+    ) -> Option<Variant> {
+        let dictionary = dictionary.to_variant();
+        let mut valid = false;
+
+        let success = iter_fn(
+            dictionary.var_sys(),
+            next_var.var_sys(),
+            addr_of_mut!(valid) as *mut u8,
```

Comment:
true, it does appear that whenever gdextension expects a bool it is exposed as a `u8`. so this should be safe since c++ treats it like a bool and so (barring UB or similar on their end) this shouldn't be made into an invalid bool. but it (probably) doesn't hurt to check anyway.

Reply:
Why still an `as` cast then?
`addr_of_mut!` already gives you a `*mut u8`.

(Also, why does clippy not catch unnecessary conversions?)

Reply:
i dont know but the unneccessary as cast was an accident i just noticed